zio-kafka/src/main/scala/zio/kafka/consumer/internal/Runloop.scala

```diff
@@ -448,37 +454,32 @@ private[consumer] final class Runloop private (
    *   - Process all currently queued commands before polling instead of one by one
    *   - Immediately after polling, if there are available commands, process them instead of waiting until some periodic
    *     trigger
+   *   - Empty the commitQueue (otherwise it would fill up when there are no rebalances)
    *   - Poll only when subscribed (leads to exceptions from the Apache Kafka Consumer if not)
    *   - Poll continuously when there are (still) unfulfilled requests or pending commits
    *   - Poll periodically when we are subscribed but do not have assigned streams yet. This happens after
    *     initialization and rebalancing
    */
-  def run: ZIO[Scope, Throwable, Any] = {
-    def logPollStart(state: State): UIO[Unit] =
-      ZIO
-        .logTrace(
-          s"Starting poll with ${state.pendingRequests.size} pending requests and ${state.pendingCommits.size} pending commits"
-        )
-
+  def run: ZIO[Scope, Throwable, Any] =
     ZStream
       .fromQueue(commandQueue)
       .timeoutFail[Throwable](RunloopTimeout)(runloopTimeout)
       .takeWhile(_ != StopRunloop)
       .runFoldChunksDiscardZIO(State.initial) { (state, commands) =>
         for {
-          _            <- ZIO.logTrace(s"Processing ${commands.size} commands: ${commands.mkString(",")}")
-          updatedState <- ZIO.foldLeft(commands)(state)(handleCommand)
+          commits <- commitQueue.takeAll
+          _ <- ZIO.logTrace(s"Processing ${commits.size} commits, ${commands.size} commands: ${commands.mkString(",")}")
+          stateAfterCommits  <- handleCommits(state, commits)
```

Comment:
Let's avoid triggering all the `handleCommits` if there are no commits
```suggestion
          stateAfterCommits  <- if (commits.isEmpty) ZIO.succeed(state) else handleCommits(state, commits)
```

Reply:
No this is not good, in fact this is explicitly necessary. By doing this we enforce a poll to the broker so that previously send commits can complete. Please also see the comments around this code.

Reply:
> By doing this we enforce a poll to the broker so that previously send commits can complete.

I'm not sure to see where this is enforced 🤔

If we look at `handleCommits`,

First, it computes these variables:
```
(offsets, callback, onFailure) = asyncCommitParameters(commits)
```
where:
 - `offsets` will be an empty list
 - `callback` will have a `onSuccess` that does nothing and a `onFailure` that requeues an empty list of commits
 - `onFailure` is the same as the one in the callback
 
Then, it computes the new state by adding these commits to the state, but because the commits is empty, we just re-allocating a new state with the same data inside.

Finally, it calls `Consumer::commitAsync` with an empty list of commits.
If we dig into the code of the `KafkaConsumer`, it seems that calling `commitAsync` will do nothing else than triggering our `callback.onSuccess`, which is a noOp:
Step 1:
<img width="1162" alt="Screenshot 2023-04-17 at 12 51 04 pm" src="https://user-images.githubusercontent.com/1193670/232464116-00018814-28e5-440f-b81d-3059a24ef6b0.png">
Step 2:
<img width="1157" alt="Screenshot 2023-04-17 at 12 51 23 pm" src="https://user-images.githubusercontent.com/1193670/232464193-8bae0e3b-7921-4851-b764-b25dfff09c0d.png">
Step 3:
<img width="859" alt="Screenshot 2023-04-17 at 12 51 34 pm" src="https://user-images.githubusercontent.com/1193670/232464233-f1d6bc92-1169-4649-90fd-087333717a13.png">
Step 4:
<img width="492" alt="Screenshot 2023-04-17 at 12 53 47 pm" src="https://user-images.githubusercontent.com/1193670/232464319-40387f9c-56db-458a-a518-6358c6c61956.png">


So, I'm still unsure why we need to call `handleCommits` with an empty list of commits.

🤔

Reply:
Hum ok I think I see.
It's because of this `client.pollNoWakeup` call in `commitOffsetsAsync`:
<img width="804" alt="image" src="https://user-images.githubusercontent.com/1193670/232469106-c3a29ae0-5892-4078-9773-360b59b00014.png">
?

🤔

Reply:
Yeah, it took me long time as well. But indeed, yes, `client.pollNoWakeup` is the place where the broker is polled and callbacks are invoked.